### ISSUE TITLE
Batchnorm NAN in amp autocast mode.

### ISSUE THREAD

**Issue by kellenf** [pytorch/pytorch#115500]
### 🐛 Describe the bug

```python
import torch.nn as nn
import torch
from torch.cuda.amp import autocast

bn_layer = nn.BatchNorm1d(10).cuda()
inp = torch.rand(2, 10).cuda() * 65535 * 2
conv_layer = nn.Linear(10, 10).cuda()

with autocast(enabled=True):
    x = conv_layer(inp)
    oup = bn_layer(x)
    print('bn_running_mean', bn_layer.running_mean, 'running_mean_dtype', bn_layer.running_mean.dtype)
```
Output:
```python
bn_running_mean tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:0') running_mean_dtype torch.float32
```
And force fp32 in bn layer is not useful?
```python
with autocast(enabled=True):
    x = conv_layer(inp)
    with autocast(enabled=False, dtype=torch.float32):
        oup = bn_layer(x.float())
        print('bn_running_mean', bn_layer.running_mean, 'running_mean_dtype', bn_layer.running_mean.dtype)
```
The output:
```
bn_running_mean tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan], device='cuda:0') running_mean_dtype torch.float32
```

It seems batchnorm compute mean var in fp16? It's very easy occur overlap in amp_autocast? It's my code:
Many people meet LOSS NAN in autocast mode, maybe it's the main reason. But why mean, var is calculated in fp16?
@mingfeima  @ptrblck  Can you help me? Thanks!

### Versions

Setting:
- machine: ubuntu18.04
- pytorch version: torch1.12.1_cu113 (torch2.0.1_cu118 also can reproduce this problem

**Comment by uygnef** [comment]
any update? same issue
